object_store_factory/src/aws.rs

```diff
@@ -0,0 +1,322 @@
+            region: map.get("region").map(|s| s.to_string()),
+            access_key_id: map.get("access_key_id").map(|s| s.to_string()),
+            secret_access_key: map.get("secret_access_key").map(|s| s.to_string()),
+            session_token: map.get("session_token").map(|s| s.to_string()),
+            endpoint: map.get("endpoint").map(|s| s.to_string()),
+            bucket: map.get("bucket").unwrap().clone(),
+            _prefix: map.get("prefix").map(|s| s.to_string()),
+        })
+    }
+}
+
+pub fn build_amazon_s3_from_config(
+    config: &S3Config,
+) -> Result<Arc<dyn ObjectStore>, object_store::Error> {
+    let mut builder = AmazonS3Builder::new()
+        .with_region(config.region.clone().unwrap_or_default())
+        .with_bucket_name(config.bucket.clone())
+        .with_allow_http(true);
+
+    if let Some(endpoint) = &config.endpoint {
+        builder = builder.with_endpoint(endpoint.clone());
+    }
+
+    if let (Some(access_key_id), Some(secret_access_key)) =
+        (&config.access_key_id, &config.secret_access_key)
+    {
+        builder = builder
+            .with_access_key_id(access_key_id.clone())
+            .with_secret_access_key(secret_access_key.clone());
+
+        if let Some(token) = &config.session_token {
+            builder = builder.with_token(token.clone())
+        }
+    } else {
+        builder = builder.with_skip_signature(true)
+    }
+
+    let store = builder.build()?;
+    Ok(Arc::new(store))
+}
+
+pub fn map_options_into_amazon_s3_config_keys(
+    input_options: HashMap<String, String>,
+) -> Result<HashMap<AmazonS3ConfigKey, String>, object_store::Error> {
+    let mut mapped_keys = HashMap::new();
+
+    for (key, value) in input_options {
+        match AmazonS3ConfigKey::from_str(&key) {
+            Ok(config_key) => {
+                mapped_keys.insert(config_key, value);
+            }
+            Err(err) => {
+                return Err(err);
+            }
+        }
+    }
+
+    Ok(mapped_keys)
+}
+
+pub async fn add_amazon_s3_specific_options(
+    url: &Url,
+    options: &mut HashMap<AmazonS3ConfigKey, String>,
+) {
+    if !options.contains_key(&AmazonS3ConfigKey::Bucket)
+        && !options.contains_key(&AmazonS3ConfigKey::Endpoint)
+    {
+        let region = detect_region(url).await.unwrap();
+        options.insert(AmazonS3ConfigKey::Region, region.to_string());
+    }
+}
+
+pub fn add_amazon_s3_environment_variables(
+    options: &mut HashMap<AmazonS3ConfigKey, String>,
+) {
+    let env_vars = &[
+        ("AWS_ACCESS_KEY_ID", AmazonS3ConfigKey::AccessKeyId),
+        ("AWS_SECRET_ACCESS_KEY", AmazonS3ConfigKey::SecretAccessKey),
+        ("AWS_DEFAULT_REGION", AmazonS3ConfigKey::Region),
+        ("AWS_ENDPOINT", AmazonS3ConfigKey::Endpoint),
+        ("AWS_SESSION_TOKEN", AmazonS3ConfigKey::Token),
+        (
+            "AWS_CONTAINER_CREDENTIALS_RELATIVE_URI",
+            AmazonS3ConfigKey::ContainerCredentialsRelativeUri,
+        ),
```

Comment:
Seems like the first elements are redundant, since you can get them from the second element(s): `config_key.as_ref().to_uppercase()`

Reply:
I changed it completely to be more robust and accept more variables. I implemented something similar to their `from_env()` algorithm like here https://docs.rs/object_store/latest/src/object_store/aws/builder.rs.html#429-443.

Reply:
Nice!